### DISCUSSION TITLE
Using fixtures.FixtureRequest in plugin code?

### DISCUSSION THREAD

**Discussion by Erotemic** [pytest-dev/pytest#8512] [Q&A]
I have a pytest plugin for the xdoctest package: [xdoctest.plugin](https://github.com/Erotemic/xdoctest/blob/master/xdoctest/plugin.py)

When I wrote this code I followed the code in [_pytest.doctest](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/doctest.py) very closely.

Recently when I've been running it I've been getting a warning: 

```
  /home/joncrall/code/xdoctest/xdoctest/plugin.py:281: PytestDeprecationWarning: A private pytest class or function was used.
    fixture_request = fixtures.FixtureRequest(xdoctest_item)
```

Which seems to be because I'm not calling `FixtureRequest` with `_ispytest=True`. If I add in that line the warning goes away, but I'd rather not do that because my plugin is not part of pytest itself. However, I'm not sure what the alternative or new preferred way to call that functionality would be.

How should external plugins handle writing something like `_setup_fixtures` on a class that inherits from `pytest.Item`?

**Comment by bluetech** [comment]
Hi @Erotemic, good question!

We've added the `_ispytest` stuff to delineate private pytest API from public API. Historically it has been quite muddled.

For now, constructing `FixtureRequest` is considered private. This means that we reserve the right to break it in a minor release, without deprecation.

Since you are using it, you have two options:

1. Set `_is_pytest=True`, but that means you have to be prepared to fix the usage if it breaks and you explicitly "opt in" to it. You might want to consider running CI occasionally against pytest `main` branch to catch breakage before pytest release

   Just to be clear, most stuff is usually quite stable and doesn't break that often, it just means that it *might*.

2. Consider what it takes to make the necessary usage a public API. This probably means opening an issue with the plugin's use case and how it's currently solved, then we may discuss possible hooks or APIs to achieve it. (Your last question starts doing that, if so we can move this discussion to an issue)

**Reply by Erotemic** [reply]
Thanks, that gives me some confidence in setting `_is_pytest` will be stable enough in the near future. 

I should probably work towards item 2, but I think I first need to better understand what that fixture code is doing in the first place ,or what the purpose of it is in the builtin doctest pytest plugin (as previously stated, because my plugin is used similarly to Python's `doctest` I copied most of the pytest plugin code without fully understanding it. (To be clear I understand most of it, but I've never really been able to wrap my head around fixtures. I usually avoid them). So I think me taking time to understand fixtures better is going to be a prereq before I can ask the right questions for item 2.

**Reply by RonnyPfannschmidt** [reply]
i believe it boils down to fixture support for items instead of just python items

**Reply by Erotemic** [reply]
@RonnyPfannschmidt Can you expand on that -- specifically, for an audience (me) that might have tried to use a fixture before but didn't 100% get it?

**Reply by RonnyPfannschmidt** [reply]
@Erotemic basically right now fixtures are supported only in the python item classes for python functions, with some hack for the builtin doctest

we should take a look at having all test items support fixtures
